### ISSUE TITLE
tf.keras.ops.image.affine_transform returns incorrect image gradient with nearest-neighbor sampling

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#127241]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13.5

### Current behavior?

`tf.keras.ops.image.affine_transform` returns an incorrect first derivative with respect to image values.

In the reproducer below, the transform is fixed and differentiation is only with respect to the scalar used to construct the image values. With nearest-neighbor sampling and a fixed transform, the selected output is linear in those image values.

For the given transform and output selection, the scalar output is `(2 * t - 1) + 3 * (5 * t - 4)`, so the derivative with respect to `t` should be `17`.

The forward value is computed correctly as `-863.0`, but reverse-mode autodiff returns `30.0` for the derivative. This suggests that the backward pass is accumulating gradients from the wrong image positions or with incorrect sampling semantics.

Expected behavior?

The gradient of the selected `tf.keras.ops.image.affine_transform` output with respect to `t` should be close to `17.0`, not `30.0`, when using a fixed transform and `interpolation="nearest"`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    image = tf.reshape(
        tf.stack([(i + 1) * t - i for i in range(6)]),
        [2, 3, 1],
    )

    transform = tf.constant(
        [1, 0, 1, 0, 1, 0, 0, 0],
        dtype=t.dtype,
    )

    y = tf.keras.ops.image.affine_transform(
        image,
        transform,
        interpolation="nearest",
        fill_mode="nearest",
        fill_value=0,
        data_format="channels_last",
    )

    return y[0, 0, 0] + 3 * y[1, 0, 0]

x = tf.constant(-50.0, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 17.0

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.ops.image.affine_transform returns an incorrect image gradient")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: -863.0
actual: 30.0
expected: 17.0
BUG REPRODUCED: tf.keras.ops.image.affine_transform returns an incorrect image gradient
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun, Thank you for reporting this issue!

The `tf.keras.ops.image.affine_transform` appears to return an incorrect gradient when using interpolation='nearest'. In the provided reproduction, the forward pass produces the expected result, but the gradient with respect to `t` is `30.0` instead of the expected `17.0`.

We’ll investigate the issue on our end and provide an update. Thanks!

**Comment by RiccardoSquarcialupi** [comment]
Hi @ALinrunrun, thanks for the clean repro, made this easy to track down.

The registered gradient doesn't actually compute a proper adjoint -- it re-runs the forward resampling op on the incoming gradient through the inverted transform. That works fine most of the time, but breaks once fill_mode clamps or folds an out-of-range coordinate, which is exactly what happens here.

In your repro, the forward pass reads output column 2 from input column 3 (out of range), clamped back to column 2 by fill_mode="nearest". The gradient's inverse map hits the same clamp in reverse and ends up reading column 0 twice, so `1*2 + 3*5 = 17` becomes `30`.

Opened a fix here: #127618. It replaces that with a real scatter-based adjoint. I tested it against your exact repro (comes back 17.0) plus a broader check across fill_modes/interpolations/transforms against the real op, so I'm fairly confident in it, but happy for someone to take a closer look.

**Comment by Venkat6871** [comment]
Hi @ALinrunrun ,
Thank you for raising your concern here.
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/127618). Once it is merged, this issue will be resolved.

Hi @RiccardoSquarcialupi  ,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.

**Comment by OssAgenticsT** [comment]
Closing this issue as resolved. Feel free to open a new issue or comment here if you need further assistance!
